packages/actor-query-operation-bgp-left-deep-reordering/lib/ActorQueryOperationBgpLeftDeepReordering.ts

```diff
@@ -0,0 +1,196 @@
+import {
+  ActorQueryOperation, ActorQueryOperationTypedMediated,
+  IActorQueryOperationOutputBindings, IActorQueryOperationTypedMediatedArgs,
+} from "@comunica/bus-query-operation";
+import {Bindings} from "@comunica/bus-query-operation";
+import {IActorTest} from "@comunica/core";
+import {MultiTransformIterator} from "asynciterator";
+import {PromiseProxyIterator} from "asynciterator-promiseproxy";
+import * as RDF from "rdf-js";
+import {termToString} from "rdf-string";
+import {getTerms} from "rdf-terms";
+import {Algebra, Factory} from "sparqlalgebrajs";
+
+/**
+ * A Bgp actor whose algorithm is based on the original TPF algorithm.
+ */
+export class ActorQueryOperationBgpLeftDeepReordering extends ActorQueryOperationTypedMediated<Algebra.Bgp> {
+
+  private factory: Factory;
+
+  constructor(args: IActorQueryOperationTypedMediatedArgs) {
+    super(args, 'bgp');
+    this.factory = new Factory();
+  }
+
+  /**
+   * Finds connected subpatterns within the possibly disconnected graph pattern.
+   * @param {Quad[]} patterns
+   * @returns {Quad[][]} - Every list is a connected group of patterns.
+   */
+  public static findConnectedPatterns(patterns: Algebra.Pattern[]): IPatternCluster[] {
+
+    // Initially consider all individual triple patterns as disconnected clusters
+    let clusters: IPatternCluster[] = patterns.map((quad) => {
+      return {
+        quads:  [quad],
+        variables: getTerms(quad).filter((term) => term.termType === 'BlankNode' || term.termType === 'Variable'),
+      };
+    });
+
+    // Zero or single-triple patterns have exactly one cluster
+    if (clusters.length <= 1) {
+      return clusters;
+    }
+
+    let commonVar: RDF.Variable;
+    // Continue clustering as long as different clusters have common variables
+    do {
+      // Find a variable that occurs in more than one subpattern
+      const allVariables = Array.prototype.concat(...clusters.map((cluster) => cluster.variables));
```

Comment:
That's a nice shorthand, I'll have to remember that :-)

Reply:
(Totally stole it from stackoverflow somewhere).